tests/tram_track_graph_transformer/test_tram_track_graph_transformer.py

```diff
@@ -0,0 +1,184 @@
+    ):
+        # Arrange
+        transformer = TramTrackGraphTransformer(
+            tram_stops_and_tracks_overpass_query_result, krakow_city_configuration
+        )
+        graph = transformer.densify_graph_by_max_distance(25.0)
+
+        for node_id in osm_tram_stops.get_node_ids():
+
+            # Assert
+            assert graph.has_node(node_id)
+
+    @pytest.mark.parametrize(
+        "max_densification_distance", CORRECT_MAX_DENSIFICATION_DISTANCES
+    )
+    def test_tram_track_transformer_densify_max_distance(
+        self,
+        max_densification_distance: float,
+        tram_stops_and_tracks_overpass_query_result,
+        krakow_city_configuration,
+    ):
+        # Arrange
+        transformer = TramTrackGraphTransformer(
+            tram_stops_and_tracks_overpass_query_result, krakow_city_configuration
+        )
+        perm_nodes = transformer.permament_nodes
+        tolerance = max_densification_distance * 0.001
+
+        # Act
+        densified_graph = transformer.densify_graph_by_max_distance(
+            max_densification_distance
+        )
+
+        for perm_node in perm_nodes:
+            prev_nodes: deque[Node] = deque()
+            prev_nodes.append(perm_node)
+            visited = set()
+
+            while prev_nodes:
+                prev_node = prev_nodes.popleft()
+                visited.add(prev_node)
+
+                for next_node in densified_graph.successors(prev_node):
+                    if next_node not in perm_nodes and next_node not in visited:
+                        prev_nodes.append(next_node)
+
+                    _, _, distance = self._geod.inv(
+                        prev_node.lon, prev_node.lat, next_node.lon, next_node.lat
+                    )
+
+                    # Assert
+                    assert distance <= (max_densification_distance + tolerance)
+
+    @pytest.mark.parametrize(
+        "max_densification_distance", CORRECT_MAX_DENSIFICATION_DISTANCES
+    )
+    def test_tram_track_transformer_densify_even_spacing(
+        self,
+        max_densification_distance: float,
+        tram_stops_and_tracks_overpass_query_result,
+        krakow_city_configuration,
+    ):
+        # Arrange
+        transformer = TramTrackGraphTransformer(
+            tram_stops_and_tracks_overpass_query_result, krakow_city_configuration
+        )
+        perm_nodes = transformer.permament_nodes
+        m = 0.05
+
+        # Act
+        densified_graph = transformer.densify_graph_by_max_distance(
+            max_densification_distance
+        )
+
+        for perm_node in perm_nodes:
+            succ_nodes = densified_graph.successors(perm_node)
+
+            for succ_node in succ_nodes:
+                if succ_node in perm_nodes:
+                    continue
+                distances = []
+                prev_node = perm_node
+                next_node = succ_node
+                visited = set()
+                visited.add(prev_node)
```

Comment:
You can initialize a set like this:
```suggestion
                visited = {prev_node}
```